### PR TITLE
Issue 2114 (main)

### PR DESCRIPTION
**Pull request by JustinKyleJames**
This is a change to reset the thread pool when uploading a large file from a local cache file.  

Notes:

1. The commit for issue 2115 is simply moving some tests to a new class so that they don't get executed in every scenario.  The contents of those tests are unaltered.
2. Before release there will be a new version of libs3 and the CMakeLists.txt for this will need to be updated.  Since I don't know the final hash for that libs3 I can't make the update now.

### CODE REVIEW

File: s3/s3_transport/include/s3_transport.hpp
Comment (korydraughn) [inline s3/s3_transport/include/s3_transport.hpp:1091]: This seems costly.

How often do you expect a new thread pool to be instantiated?

Comment (JustinKyleJames) [reply s3/s3_transport/include/s3_transport.hpp:1091]: Rarely.  For example, with the defaults of 10 threads and 5 GB maximum part size, for a 100 GB file it would get re-instantiated one time (2 total instantiations).  Whatever time it takes to re-instantiate it would only be a fraction compared to the time to actually send the bytes.

Comment (korydraughn) [reply s3/s3_transport/include/s3_transport.hpp:1091]: I feel much better about the change now.

Please mark this conversation as resolved.

**Review by korydraughn — COMMENTED** [review]
Changes look good.

Do the tests pass?

**Comment by JustinKyleJames** [comment]
> Changes look good.
> 
> Do the tests pass?

Yes, they did.

**Review by korydraughn — APPROVED** [review]
Let's wait for one more approval before adding the pounds.

File: packaging/resource_suite_s3_nocache.py
Comment (alanking) [inline packaging/resource_suite_s3_nocache.py:2358]: Consider using `self.assertTrue` instead of naked `assert`.

**Review by alanking — APPROVED** [review]
One minor/optional thing. Up to you

File: packaging/resource_suite_s3_nocache.py
Comment (JustinKyleJames) [reply packaging/resource_suite_s3_nocache.py]: done

**Review by alanking — COMMENTED** [review]
Nice. Please squash and add #s

**Review by korydraughn — APPROVED** [review]

**Comment by JustinKyleJames** [comment]
> Nice. Please squash and add #s

Done

**Pull request merged**
